contributor-guide/modules/ROOT/pages/testing/intro.adoc

```diff
@@ -1,29 +1,95 @@
+** GCC (GNU Compiler Collection)
+** Clang
+** MSVC (Microsoft Visual pass:[C++])
+** ICC (Intel pass:[C++] Compiler)
++
+The Boost user base can be using older versions of these compilers, so strive for compatibility with a reasonable range of versions if possible.
++
+If your library depends on other Boost libraries or external libraries, ensure they are compatible with the compilers you are targeting. Be clear about any dependencies or prerequisites in your documentation.
++
+Be wary of non-standard compiler features or extensions. If you must use them, guard them with appropriate preprocessor checks.
+Boost provides its own set of configuration macros to help with this.
+
+. Compatibility with various versions of the pass:[C++] Standard: pass:[C++]11, pass:[C++]14, pass:[C++]17, pass:[C++]20, and pass:[C++]23. Some Boost libraries support many of the standards, while others target only more recent ones. 
++
+Note:: Supporting pass:[C++]03 is no longer considered good practice.
+
+. Different operating systems, including Linux, Windows, macOS, and others like various BSDs.
+
+. Different architectures: x86, x64, ARM, MIPS. Architecture can affect word size (usually 32 or 64 bit), endianness, memory alignment, inline assembly, cache sizes, latency, and other memory and performance issues.
+
+When you have outlined your test matrix, study the predefined macros available to assist you, and make adjustments to your matrix appropriately.
+
+== Use Predefined Macros from Boost.Config
+
+There are a set of macros in the boost:config[] library that can be used to identify compiler capabilities, platform specifics, and other configuration details. These macros are designed to smooth out differences between compilers and platforms, allowing for more portable code, for example:
+
+=== Compiler Identification
+
+* `BOOST_COMP_GNUC`: Defined if the compiler is GCC.
+* `BOOST_COMP_MSVC`: Defined if the compiler is Microsoft Visual C++.
+
+=== Compiler Version Checks
+
+* `BOOST_COMP_MSVC >= BOOST_VERSION_NUMBER(19, 0, 0)`: Checks if the MSVC version is 19 or greater.
+
+=== Platform Checks
+
+* `BOOST_OS_LINUX`: Defined if the operating system is Linux.
+* `BOOST_ARCH_X86_64`: Defined if the architecture is x86_64.
+
+=== Standard Library Checks
+
+* `BOOST_LIB_C_GNU`: Defined if the C standard library is from https://gcc.gnu.org/onlinedocs/libstdc++/[GNU].
+* `BOOST_LIB_STD_DINKUMWARE`: Defined if the standard library is from https://docwiki.embarcadero.com/RADStudio/Sydney/en/Dinkumware_Standard_C%2B%2B_Library[Dinkumware] (often associated with MSVC).
+
+=== Feature Checks
+
+* `BOOST_NO_CXX14_RETURN_TYPE_DEDUCTION`: Defined if the compiler does not support return type deduction introduced in C++14.
+* `BOOST_NO_CXX11_AUTO_DECLARATIONS`: Defined if the compiler does not support auto declarations from C++11. For example:
++
+[source,cpp]
+----
+#include <boost/config.hpp>
+
+#ifdef BOOST_NO_CXX11_AUTO_DECLARATIONS
```

Comment:
This is invalid as a complete source file because these are expressions, which would need an enclosing declaration.  You can include `// ...` in some line before the `#ifdef` to indicate this is just an example and that what comes after `// ...` is just supposed to be somewhere in the file.